### ISSUE TITLE
Invalid type of return parameter in bean generated for custom HTTP client

### ISSUE THREAD

**Issue by l0co** [micronaut-projects/micronaut-core#4592]
In `2.2.0`, works in `2.1.4`.

I have a custom HTTP client created with factory and wrapped into around advice:

```java
@Factory
public class SafeLowLevelClientFactory {

	public static final String INTERNAL_SERVICE_CLIENT = "lowLevelInternalServiceClient";

	@Bean @Named(INTERNAL_SERVICE_CLIENT)
	@ConvertException(ServiceUnavailableExceptionConverter.class)
	@CircuitBreaker(includes = ServiceUnavailableException.class)
	public SafeHttpClient createInternalServiceClient(LoadBalancerResolver loadBalancerResolver, HttpClientConfiguration configuration) {
		return new DefaultSafeHttpClient(loadBalancerResolver.resolve("safe-non-existing-service").orElseThrow(),
			configuration, "/safe-non-existing-path");
	}

}
```

`@ConvertException` is my around advice which basically tries to convert a group of exceptions to some other execption, with this implementation:

```java
	@Override
	public Object intercept(MethodInvocationContext<Object, Object> context) {
		Class<? extends IExceptionConverter>[] converterClasses = context
			.findAnnotation(ConvertException.class).orElseThrow()
			.getValue(Class[].class).orElseThrow();

		try {

			Object res = context.proceed();

			if (Publishers.isConvertibleToPublisher(context.getReturnType().getType())) {
				Flowable<?> observable = ConversionService.SHARED.convert(res, Flowable.class)
					.orElseThrow(() -> new IllegalStateException("Unconvertible Reactive type: " + res));
				return observable.onErrorResumeNext(ex -> {
					try {
						support(ex, converterClasses);
						return Flowable.error(ex);
					} catch (Exception e) {
						return Flowable.error(e);
					}
				});
			}

			return res;

		} catch (Throwable t) {

			support(t, converterClasses);

			throw new RuntimeException(t);
		}
	}
```

This was covered by test which was working to `2.1.4` but in `2.2.0` I have the following exception from this code during accessing `client.retrieve("/").blockingSingle()`:

```
Caused by: java.lang.ClassCastException: class io.micronaut.core.async.publisher.Publishers$JustThrowPublisher cannot be cast to class io.reactivex.Flowable (io.micronaut.core.async.publisher.Publishers$JustThrowPublisher and io.reactivex.Flowable are in unnamed module of loader 'app')
	at io.luna.feature.core.test.client.$SafeLowLevelClientFactory$CreateInternalServiceClient3Definition$Intercepted.retrieve(Unknown Source)
	at io.luna.feature.core.test.client.SafeLowLevelClientTest.lambda$testCircucitBreakerLowLevelSafeClient$0(SafeLowLevelClientTest.java:73)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:55)
	... 105 more
```

I'm also attaching the screen with breakpoint on this exception:

![image](https://user-images.githubusercontent.com/1880196/99963965-deb7e180-2d92-11eb-8f0c-b5d0a814962e.png)

**Comment by jameskleeh** [comment]
Can you upload a sample app? Hard to say what the issue might be without debugging it. Do you know what call to `Publishers.just(Throwable)` is creating the result?

Also getting the values from the annotation can be simplified:

`context.classValues(ConvertException.class).orElseThrow()`

**Comment by l0co** [comment]
I've extracted the code [here](https://github.com/l0co/micronaut-playground/tree/micronaut-4592). You can run `SafeLowLevelClientTest`. I cannot reproduce the problem, though, while it is still present in my real project when I use Micronaut 2.2.0.

But, I have some more insight now, after fourth throwing of `ServiceUnavailableException`, `DefaultRetryInterceptor` gives up and should forward error upwards. In my real project it ends up here (marked with a comment - in catch clause):

```java
@Singleton
public class DefaultRetryInterceptor implements MethodInterceptor<Object, Object> {

    @Override
    public Object intercept(MethodInvocationContext<Object, Object> context) {
        Optional<AnnotationValue<Retryable>> opt = context.findAnnotation(Retryable.class);
        // ....
    
        InterceptedMethod interceptedMethod = InterceptedMethod.of(context);
        try {
            retryState.open();
            // Retry method call before we have actual Publisher/CompletionStage result
            Object result = retrySync(context, retryState, interceptedMethod);
            switch (interceptedMethod.resultType()) {
                case PUBLISHER:
                    Flowable<Object> flowable = Flowable.fromPublisher((Publisher<?>) result);
                    return interceptedMethod.handleResult(
                            flowable.onErrorResumeNext(retryFlowable(context, retryState, flowable))
                                    .doOnNext(o -> retryState.close(null))
                    );
                case COMPLETION_STAGE:
                    CompletableFuture<Object> newFuture = new CompletableFuture<>();
                    Supplier<CompletionStage<?>> retrySupplier = () -> interceptedMethod.interceptResultAsCompletionStage(this);
                    ((CompletionStage<?>) result).whenComplete(retryCompletable(context, retryState, newFuture, retrySupplier));
                    return interceptedMethod.handleResult(newFuture);
                case SYNCHRONOUS:
                    retryState.close(null);
                    return result;
                default:
                    return interceptedMethod.unsupported();
            }
        } catch (Exception e) {
            return interceptedMethod.handleException(e); // <--------------------- here it ends
        }
    }

}
```

But in the test project this line is never reached. The thing is I cannot compare what is the difference between both executions because of undebuggable reactive code... The question is then, what can happen in the code to end up in this line?

And, if you can reproduce the problem to reach this line, and the `interceptedMethod` will be of `PublisherInterceptedMethod`, this code will finally return `JustThrowPublisher`, while `RxHttpClient.retrieve()` should return `Flowable`. `JustThrowPublisher` is not `Flowable` and here comes the error from...

**Comment by l0co** [comment]
OK, I added some more lines of code to the test and now this problem is reproducible in my example repo :)

This happens only when a circuit is opened and the exception should be thrown immediately without even calling the method. Generally it looks that for Micronaut `2.1.4` the error was published by the publisher, while in `2.2.0` it's thrown.

But I suppose there's also a second bug in `DefaultRetryInterceptor` which forwards such caught error using `JustThrowPublisher` which is not compatible with `Flowable` return type.

**Comment by jameskleeh** [comment]
The issue here is that the executable method represents the super interface HttpClient instead of RxHttpClient. A conversion is done from the just throw publisher but the executable method represents one that returns a publisher, so no conversion is necessary.

**Comment by l0co** [comment]
Looks fixed in `2.2.1`.
